File: Arrays/Hash and Maps/2461. Maximum Sum of Distinct Subarrays With Length K.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maximumSubarraySum = function(nums, k) 
{
    let begin = 0;
    let end = 0;
    let currentSum = 0;
    let answer = 0;
    let numIndex = new Map();

    while(end < nums.length)
    {
        let currentNum = nums[end];
        let lastOccurance = numIndex.has(currentNum) ? numIndex.get(currentNum) : -1;

        while(begin <= lastOccurance || end - begin + 1 > k)
        {
            currentSum -= nums[begin];
            begin += 1;
        }

        numIndex.set(currentNum, end);
        currentSum += currentNum;

        if(end - begin + 1 === k)
            answer = Math.max(answer, currentSum);
        
        end += 1;
    }

    return answer;
};

File: Arrays/Sliding Window/2461. Maximum Sum of Distinct Subarrays With Length K.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        begin = 0
        answer = 0
        end = 0
        current_sum = 0
        number_index = {}

        while end < len(nums):
            current_number = nums[end]
            last_occurance = number_index.get(current_number, -1)

            #check if the number already exisited in window or window size is exceeding
            while begin <= last_occurance or end - begin + 1 > k:
                current_sum -= nums[begin]
                begin += 1

            number_index[current_number] = end
            current_sum += current_number
            if end - begin + 1 == k:
                answer = max(answer, current_sum)
            
            end += 1
        
        return answer
